File: extension/webview/src/component/frames/filter.ts
class Parser {
    input: string;
    pos: number;
    constructor(input) {
        this.input = input;
        this.pos = 0;
    }

    // 解析表达式
    parse() {
        return this.parseExpression();
    }

    // 解析表达式 (包含 && 和 ||)
    parseExpression() {
        let expr = this.parseTerm();

        while (this.peek() === '&&' || this.peek() === '||') {
            let operator = this.consume(undefined);
            let right = this.parseTerm();
            expr = { type: 'Binary', left: expr, operator: operator, right: right };
        }

        return expr;
    }

    // 解析单个term (处理括号或比较操作)
    parseTerm(): any {
        if (this.peek() === '(') {
            this.consume(undefined);  // 消耗 '('
            let expr = this.parseExpression();
            if (this.peek() !== ')') {
                throw new Error('Missing closing parenthesis');
            }
            this.consume(undefined);  // 消耗 ')'
            return { type: 'Group', expression: expr };
        } else {
            return this.parseComparison();
        }
    }

    // 解析比较操作 (例如 =, !=, <, >, >=, <=)
    parseComparison(): any {
        let left = this.parseValue();

        let operator = this.peekComparisonOperator();
        if (operator) {
            this.consume(operator);
            let right = this.parseValue();
            return { type: 'Binary', left: left, operator: operator, right: right };
        }
        // else {
        //     throw new Error("no value");
        // }

        return left;
    }

    // 解析值（例如变量名或数字）
    parseValue() {
        this.skipWhitespace();
        let value = '';
        while (this.pos < this.input.length && !this.isWhitespace(this.input[this.pos]) &&
               !this.isSpecialChar(this.input[this.pos])) {
            value += this.input[this.pos++];
        }
        this.skipWhitespace();
        if (value.length === 0) {
            throw new Error('Expected a value');
        }

        return { type: 'Value', value: value };
    }

    // 查看接下来的比较操作符
    peekComparisonOperator() {
        const ops = ['==', '!=', '>=', '<=', '>', '<'];
        for (let op of ops) {
            if (this.input.startsWith(op, this.pos)) {
                return op;
            }
        }
        return null;
    }

    // 查看下一个token (包括 &&, ||)
    peek() {
        this.skipWhitespace();
        if (this.input.startsWith('&&', this.pos)) return '&&';
        if (this.input.startsWith('||', this.pos)) return '||';
        if (this.input[this.pos] === '(') return '(';
        if (this.input[this.pos] === ')') return ')';
        return null;
    }

    // 消耗一个token
    consume(expected) {
        this.skipWhitespace();
        if (expected && !this.input.startsWith(expected, this.pos)) {
            throw new Error(`Expected "${expected}"`);
        }

        let token = expected || this.input[this.pos] + this.input[this.pos + 1];
        this.pos += token.length;
        return token;
    }

    // 跳过空白字符
    skipWhitespace() {
        while (this.pos < this.input.length && this.isWhitespace(this.input[this.pos])) {
            this.pos++;
        }
    }

    // 判断是否为空白字符
    isWhitespace(char) {
        return /\s/.test(char);
    }

    // 判断是否为特殊字符
    isSpecialChar(char) {
        return ['=', '!', '>', '<', '&', '|', '(', ')'].includes(char);
    }
}

// const input = "tcp.ip  && (act.count >= 1 || ppc.c < 12)";
// const parser = new Parser(input);
// const result = parser.parse();

// console.log(JSON.stringify(result, null, 2));

File: crates/shark/src/common/filter.rs
use std::cmp;
use std::collections::{HashMap, HashSet};
use std::str::Chars;
use std::iter::Peekable;
pub trait Superem{
    
}
pub trait FilterValue {
    
}

pub struct PacketProps {
    _map: HashMap<&'static str, HashSet<&'static str>>,
}

impl PacketProps{
    pub fn new() -> Self {
        Self{_map: HashMap::new()}
    }
    pub fn add(&mut self, key: &'static str, val: &'static str){
        if let Some(set) = self._map.get_mut(&key) {
            set.insert(val);
        } else {
            let mut _set = HashSet::new();
            _set.insert(val);
            self._map.insert(key, _set);
        }
    }
    pub fn get(&self, key: &'static str) -> Option<Vec<&'static str>>{
        if let Some(set) = self._map.get(&key) {
            let mut aa: Vec<&'static str> = set.into_iter().map(|f| *f).collect();
            aa.sort();
            return Some(aa);
        }
        None
    }
    pub fn merge(&mut self, other: &mut PacketProps) {
        for (key, values) in other._map.drain() {
            self._map
                .entry(key)
                .or_insert_with(HashSet::new)
                .extend(values);
        }
    }

    pub fn match_expr(&self, statement: &str) -> bool {
        let mut parser = Parser::new(statement);
        if let Ok(expr) = parser.parse() {
            return evaluate_expression(&expr, &self._map);
        } 
        false
    }
}

#[derive(Debug, PartialEq)]
pub enum Operator {
    And,
    Or,
    Equal,
    NotEqual,
    GreaterThan,
    LessThan,
    GreaterThanOrEqual,
    LessThanOrEqual,
}

#[derive(Debug, PartialEq)]
pub enum Expr {
    Binary(Box<Expr>, Operator, Box<Expr>),
    Group(Box<Expr>),
    Value(String),
}

pub struct Parser<'a> {
    chars: Peekable<Chars<'a>>,
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            chars: input.chars().peekable(),
        }
    }

    pub fn parse(&mut self) -> Result<Expr, String> {
        self.parse_expression()
    }

    pub fn parse_expression(&mut self) -> Result<Expr, String> {
        self.skip_whitespace();
        let mut expr = self.parse_term()?;

        while let Some(op) = self.peek_logical_operator() {
            self.consume_logical_operator();
            let right = self.parse_term()?;
            expr = Expr::Binary(Box::new(expr), op, Box::new(right));
        }

        Ok(expr)
    }

    pub fn parse_term(&mut self) -> Result<Expr, String> {
        self.skip_whitespace();
        if self.peek_char() == Some('(') {
            self.consume_char();
            let expr = self.parse_expression()?;
            self.skip_whitespace(); 
            if self.peek_char() == Some(')') {
                self.consume_char();
                Ok(Expr::Group(Box::new(expr)))
            } else {
                Err("Expected closing parenthesis".to_string())
            }
        } else {
            self.parse_comparison()
        }
    }

    fn parse_comparison(&mut self) -> Result<Expr, String> {
        self.skip_whitespace();
        let left = self.parse_value()?;

        if let Some(op) = self.peek_comparison_operator() {
            self.skip_whitespace();
            let right = self.parse_value()?;
            Ok(Expr::Binary(Box::new(left), op, Box::new(right)))
        } else {
            Ok(left)
            // Err("No value".to_string())
        }
    }

    fn parse_value(&mut self) -> Result<Expr, String> {
        self.skip_whitespace();
        let mut value = String::new();
        while let Some(c) = self.peek_char() {
            if c.is_whitespace() || c == '&' || c == '|' || c == '(' || c == ')' || c == '=' || c == '!' || c == '>' || c == '<' {
                break;
            }
            value.push(c);
            self.consume_char();
        }
        if !value.is_empty() {
            Ok(Expr::Value(value))
        } else {
            Err("Expected value".to_string())
        }
    }

    fn peek_char(&mut self) -> Option<char> {
        self.chars.peek().copied()
    }

    fn consume_char(&mut self) {
        self.chars.next();
    }

    fn peek_logical_operator(&mut self) -> Option<Operator> {
        self.skip_whitespace();
        let chars: String = self.chars.clone().take(2).collect();
        match chars.as_str() {
            "&&" => Some(Operator::And),
            "||" => Some(Operator::Or),
            _ => None,
        }
    }

    fn consume_logical_operator(&mut self) {
        self.chars.next();
        self.chars.next();
    }
    fn skip_whitespace(&mut self) {
        loop {
            if let None = self.chars.next_if_eq(&' ') {
                break;
            }
        }
    }

    fn next_operetor(&mut self) -> String {
        let mut str = "".to_string();
        loop {
            if let Some(ch) = self.chars.next_if(|c| "=!><".contains(*c)) {
                str.push(ch);
            } else {
                break;
            }
        }
        str
    }
    fn peek_comparison_operator(&mut self) -> Option<Operator> {
        self.skip_whitespace();
        let chars = self.next_operetor();
        match chars.as_str() {
            "==" => Some(Operator::Equal),
            "!=" => Some(Operator::NotEqual),
            ">=" => Some(Operator::GreaterThanOrEqual),
            "<=" => Some(Operator::LessThanOrEqual),
            ">" => Some(Operator::GreaterThan),
            "<" => Some(Operator::LessThan),
            _ => None,
        }
    }
}
// Evaluator to match expression with a dictionary of values
pub fn evaluate_expression(expr: &Expr, data: &HashMap<&'static str, HashSet<&'static str>>) -> bool {
    match expr {
        Expr::Binary(left, op, right) => {
            let left_val = evaluate_expression(left, data);
            let right_val = evaluate_expression(right, data);

            match op {
                Operator::And => left_val && right_val,
                Operator::Or => left_val || right_val,
                _ => {
                    let left_value = if let Expr::Value(ref key) = **left {
                        data.get(&(key.as_ref()))
                    } else {
                        return false;
                    };

                    let right_value:&str = if let Expr::Value(ref val) = **right {
                        val
                    } else {
                        return false;
                    };
                    if let Some(_set) = left_value {
                        for v in _set.iter() {
                            let left = *v;
                            let max_len = cmp::max(left.len(), right_value.len());
                            let result = match op {
                                Operator::Equal => left == right_value,
                                Operator::NotEqual => left != right_value,
                                Operator::GreaterThan => {
                                    let _left = format!("{:0>max_len$}", left);
                                    let _right = format!("{:0>max_len$}", right_value);
                                    _left > _right
                                },
                                Operator::LessThan => {
                                    let _left = format!("{:0>max_len$}", left);
                                    let _right = format!("{:0>max_len$}", right_value);
                                    _left < _right
                                },
                                Operator::GreaterThanOrEqual => {
                                    let _left = format!("{:0>max_len$}", left);
                                    let _right = format!("{:0>max_len$}", right_value);
                                    _left >= _right

                                },
                                Operator::LessThanOrEqual => {
                                    let _left = format!("{:0>max_len$}", left);
                                    let _right = format!("{:0>max_len$}", right_value);
                                    _left <= _right
                                },
                                _ => false,
                            };
                            if result {
                                return true;
                            }
                        }
                        return false
                    } else {
                        return false;
                    }
                }
            }
        }
        Expr::Group(inner_expr) => evaluate_expression(inner_expr, data),
        Expr::Value(key) => {
            data.contains_key(&(*key).as_ref())
        },
    }
}